link_prediction.py

```diff
@@ -0,0 +1,248 @@
+        neg_pred.append(score[num_batch:])
+
+    pos_pred = torch.cat(pos_pred, dim=0)
+    neg_pred = torch.cat(neg_pred, dim=0).view(-1, args.neg_ratio)
+
+    input_dict = {"y_pred_pos": pos_pred, "y_pred_neg": neg_pred}
+    mrr = evaluator.eval(input_dict)["mrr_list"]
+    return mrr, mrr.mean()
+
+
+def train(args, graph, features, train_set, valid_set, model):
+    total_loss = 0.0
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+    dataloader = get_dataloader(args, graph, features, train_set)
+
+    for epoch in tqdm.trange(args.epochs):
+        model.train()
+        for step, (input_features, compacted_pairs, blocks) in enumerate(
+            dataloader
+        ):
+            # Get the embeddings of the input nodes.
+            y = model(blocks, input_features)
+            logits = model.predictor(
+                y[compacted_pairs[0]] * y[compacted_pairs[1]]
+            ).squeeze()
+
+            # Construct the labels.
+            labels = torch.zeros_like(logits)
+            num_batch = compacted_pairs[0].shape[0] // (args.neg_ratio + 1)
+            labels[:num_batch] = 1.0
+
+            # Compute loss.
+            loss = F.binary_cross_entropy_with_logits(logits, labels.float())
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            total_loss += loss.item()
+            print(
+                f"Epoch {epoch:05d} | "
+                f"Step {step:05d} | "
+                f"Loss {(total_loss) / (step + 1):.4f}",
+                end="\r",
+            )
+
+        # Evaluate the model.
+        print("Validation")
+        valid_mrr = evaluate(args, graph, features, valid_set, model)
+        print(f"Valid MRR {valid_mrr.item():.4f}")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="OGBL-Citation2 (GraphBolt)")
+    parser.add_argument("--epochs", type=int, default=10)
+    parser.add_argument("--lr", type=float, default=0.0005)
+    parser.add_argument("--num-layers", type=int, default=3)
+    parser.add_argument("--neg-ratio", type=int, default=1000)
+    parser.add_argument("--batch-size", type=int, default=512)
+    parser.add_argument("--num-workers", type=int, default=4)
+    parser.add_argument(
+        "--fanout",
+        type=str,
+        default="15,10,5",
+        help="Fan-out of neighbor sampling. Default: 15,10,5",
+    )
+    parser.add_argument(
+        "--mode",
+        default="mixed",
+        choices=["cpu", "mixed", "puregpu"],
+        help="Training mode. 'cpu' for CPU training, 'mixed' for CPU-GPU mixed "
+        "training, 'puregpu' for pure-GPU training.",
+    )
+    return parser.parse_args()
+
+
+def main(args):
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    # Load and preprocess dataset.
+    print("Loading data")
+    dataset = gb.OnDiskDataset(
+        "/home/ubuntu/wklwork/work#4/example_ogbl_citation2"
```

Comment:
Hi, I would like to run the example myself. Could you please attach the YAML file and a description of how to prepare the data set for it?

Reply:
Thank you for your attention✨. This example is currently a provisional version due to potential changes in many of Graphbolt's APIs. For now, it serves as an internal reference. We plan to merge it into the master branch in the future and will provide an official channel for dataset downloading and processing. @frozenbugs

Reply:
Yes, @keli-wen once it is ready, please also draft readme so that ppl can run it by themselves. Thanks @anko-intel for bringing it up.